app/models/workers_scaling/experiment_resources_interface.rb

```diff
@@ -0,0 +1,221 @@
+        worker.simulations_left = simulations_left if worker.simulations_left.blank? or overwrite
+        worker.save
+      end
+    end
+
+    ##
+    # Stops worker after completing its current simulation execution
+    def soft_stop_worker(sm_uuid)
+      limit_worker_simulations(sm_uuid, 1)
+    end
+
+    ##
+    # Returns list of workers records for infrastructure
+    def get_workers_records_list(infrastructure, params = {})
+      get_workers_records_cursor(infrastructure, params).to_a
+    end
+
+    ##
+    # Returns workers records count for infrastructure
+    def get_workers_records_count(infrastructure, params = {})
+      get_workers_records_cursor(infrastructure, params).count
+    end
+
+    ##
+    # Returns overall Workers count for Experiment matching given params
+    def count_all_workers(params = {})
+      get_available_infrastructures
+          .flat_map { |infrastructure| get_workers_records_count(infrastructure, params) }
+          .reduce(0) { |sum, count| sum + count }
+    end
+
+    ##
+    # Returns worker record for given sm_uuid
+    def get_worker_record_by_sm_uuid(sm_uuid)
+      get_available_infrastructures.flat_map do |infrastructure|
+        get_workers_records_cursor(infrastructure, cond: {sm_uuid: sm_uuid}).first
+      end .first
+    end
+
+    ##
+    # Yields workers for given records
+    # Usage:
+    #   yield_workers(records) do |worker|
+    #      worker.<some method>
+    #   end
+    # Workers commands: restart, stop, destroy_record
+    def yield_workers(records)
+      records.each do |record|
+        get_facade_for(record.infrastructure).yield_simulation_manager(record) {|worker| yield worker}
+      end
+    end
+
+    private
+
+    ##
+    # Checks whether all fields from narrower are equal with corresponding fields from wider
+    # when exact flag is set to false. Performs full comparison when exact flag is true.
+    # By default exact flag is set to false
+    # Returns true when infrastructures are equal, false otherwise.
+    def infrastructures_equal?(narrower, wider, exact=false)
+      # TODO replace with infrastructure id
+      return false if wider[:name] != narrower[:name]
+      if exact
+        return false if wider[:params] != narrower[:params]
+      else
+        narrower[:params].each do |key, value|
+          return false if wider[:params][key] != value
+        end
+      end
+      true
+    end
+
+    ##
+    # Returns InfrastructureFacade for given infrastructure name
+    # Previously accessed facades are cached
+    def get_facade_for(infrastructure_name)
+      unless @facades_cache.has_key? infrastructure_name
+        throw NoSuchInfrastructureError unless get_enabled_infrastructures.map {|infrastructure| infrastructure[:name]}
+                                                                            .include? infrastructure_name
+        @facades_cache[infrastructure_name] = InfrastructureFacadeFactory.get_facade_for infrastructure_name
```

Comment:
@mpaciore @mwrona - Nazwa metody dubluje tą z fabryki, co nie do końca mi się podoba, bo ma w środku dodatkową logikę. Już lepiej, gdyby nazwać tą metodę inaczej (``cached_infrastructure_facade(...)``?). Cache'owanie obiektów fasad w przypadku kiedy w innym miejscu jest znacznie cięższe yieldowanie workerów (bez użycia onsite) jest przesadą, ale skoro jest, niech zostanie, tylko z inną nazwą.

Tutaj realizacją RAII (gdzie zasobami jest np. połączenie SSH) jest yieldowanie a nie tworzenie obiektu fasady, bo to Ruby.